File: 11. Container With Most Water/11. Container With Most Water.rb
# @param {Integer[]} height
# @return {Integer}
def max_area(height)
    left, right = 0, height.length - 1
    max_water = 0
    
    while left < right
        width = right - left
        h = [height[left], height[right]].min
        area = width * h
        
        max_water = [max_water, area].max
        
        if height[left] < height[right]
            left += 1
        else
            right -= 1
        end
    end
    
    max_water
end

File: 11. Container With Most Water/11. Container With Most Water.go
func maxArea(height []int) int {
    left, right := 0, len(height) - 1
    maxWater := 0
    
    for left < right {
        width := right - left
        h := min(height[left], height[right])
        area := width * h
        
        maxWater = max(maxWater, area)
        
        if height[left] < height[right] {
            left++
        } else {
            right--
        }
    }
    
    return maxWater
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}